From 4f0cafcb2850a08d9081eea3ac8bec17b9a9881c Mon Sep 17 00:00:00 2001
From: John Johansen <john.johansen@canonical.com>
Date: Sat, 9 Oct 2021 02:30:41 -0700
Subject: [PATCH 13095/15333] apparmor: refactor network sock mediation in
 preparation for inet mediation

Refactor network mediation, introducing the stub code for the fine
grained inet mediation. This is a preparatory step and does not change
mediation.

Signed-off-by: John Johansen <john.johansen@canonical.com>
---
 security/apparmor/Makefile          |  2 +-
 security/apparmor/af_inet.c         | 81 +++++++++++++++++++++++++++++
 security/apparmor/include/af_inet.h | 36 +++++++++++++
 security/apparmor/include/net.h     |  7 +--
 security/apparmor/lsm.c             | 66 +++++++++++++++++++----
 security/apparmor/net.c             | 20 ++++---
 6 files changed, 190 insertions(+), 22 deletions(-)
 create mode 100644 security/apparmor/af_inet.c
 create mode 100644 security/apparmor/include/af_inet.h

diff --git a/security/apparmor/Makefile b/security/apparmor/Makefile
index 12fb419714c0..8aed0d7ed6f5 100644
--- a/security/apparmor/Makefile
+++ b/security/apparmor/Makefile
@@ -6,7 +6,7 @@ obj-$(CONFIG_SECURITY_APPARMOR) += apparmor.o
 apparmor-y := apparmorfs.o audit.o capability.o task.o ipc.o lib.o match.o \
               path.o domain.o policy.o policy_unpack.o procattr.o lsm.o \
               resource.o secid.o file.o policy_ns.o label.o mount.o net.o \
-              policy_compat.o af_unix.o
+              policy_compat.o af_unix.o af_inet.o
 apparmor-$(CONFIG_SECURITY_APPARMOR_HASH) += crypto.o
 
 obj-$(CONFIG_SECURITY_APPARMOR_KUNIT_TEST) += apparmor_policy_unpack_test.o
diff --git a/security/apparmor/af_inet.c b/security/apparmor/af_inet.c
new file mode 100644
index 000000000000..780c905b029f
--- /dev/null
+++ b/security/apparmor/af_inet.c
@@ -0,0 +1,81 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+/*
+ * AppArmor security module
+ *
+ * This file contains AppArmor inet fine grained mediation
+ *
+ * Copyright 2024 Canonical Ltd.
+ *
+ */
+
+#include <net/tcp_states.h>
+
+#include "include/audit.h"
+#include "include/af_inet.h"
+#include "include/apparmor.h"
+#include "include/file.h"
+#include "include/label.h"
+#include "include/net.h"
+#include "include/path.h"
+#include "include/policy.h"
+#include "include/cred.h"
+
+
+/* no kernel_t early bailout */
+/* NOTE: already lifted label_mediates into lsm.c */
+int aa_inet_create_perm(struct aa_label *label, int family, int type,
+			int protocol)
+{
+	return aa_af_perm(current_cred(), label, OP_CREATE,
+			  AA_MAY_CREATE, family, type,
+			  protocol);
+}
+
+int aa_inet_bind_perm(struct socket *sock, struct sockaddr *addr,
+		      int addrlen)
+{
+	return aa_sk_perm(OP_BIND, AA_MAY_BIND, sock->sk);
+}
+
+int aa_inet_connect_perm(struct socket *sock, struct sockaddr *addr,
+			 int addrlen)
+{
+	return aa_sk_perm(OP_CONNECT, AA_MAY_CONNECT, sock->sk);
+}
+
+int aa_inet_listen_perm(struct socket *sock, int backlog)
+{
+	return aa_sk_perm(OP_LISTEN, AA_MAY_LISTEN, sock->sk);
+}
+
+/* ability of sock to connect, not peer address binding */
+int aa_inet_accept_perm(struct socket *sock, struct socket *newsock)
+{
+	return aa_sk_perm(OP_ACCEPT, AA_MAY_ACCEPT, sock->sk);
+}
+
+/* sendmsg, recvmsg. */
+int aa_inet_msg_perm(const char *op, u32 request, struct socket *sock,
+		     struct msghdr *msg, int size)
+{
+	return aa_sk_perm(op, request, sock->sk);
+}
+
+/* getopt, setopt */
+int aa_inet_opt_perm(const char *op, u32 request, struct socket *sock,
+		     int level, int optname)
+{
+	return aa_sk_perm(op, request, sock->sk);
+}
+
+/* revaliation, get/set attr/getsockname/peername */
+int aa_inet_sock_perm(const char *op, u32 request, struct socket *sock)
+{
+	return aa_sk_perm(op, request, sock->sk);
+}
+
+int aa_inet_file_perm(const struct cred *subj_cred, struct aa_label *label,
+		      const char *op, u32 request, struct socket *sock)
+{
+	return aa_label_sk_perm(subj_cred, label, op, request, sock->sk);
+}
diff --git a/security/apparmor/include/af_inet.h b/security/apparmor/include/af_inet.h
new file mode 100644
index 000000000000..601ab44bdfe1
--- /dev/null
+++ b/security/apparmor/include/af_inet.h
@@ -0,0 +1,36 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+/*
+ * AppArmor security module
+ *
+ * This file contains AppArmor inet/inet6 fine grained mediation
+ *
+ * Copyright 2024 Canonical Ltd.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, version 2 of the
+ * License.
+ */
+#ifndef __AA_AF_INET_H
+#define __AA_AF_INET_H
+
+#include "label.h"
+
+int aa_inet_sock_perm(const char *op, u32 request, struct socket *sock);
+int aa_inet_create_perm(struct aa_label *label, int family, int type,
+			int protocol);
+int aa_inet_bind_perm(struct socket *sock, struct sockaddr *address,
+		      int addrlen);
+int aa_inet_connect_perm(struct socket *sock, struct sockaddr *address,
+			 int addrlen);
+int aa_inet_listen_perm(struct socket *sock, int backlog);
+int aa_inet_accept_perm(struct socket *sock, struct socket *newsock);
+int aa_inet_msg_perm(const char *op, u32 request, struct socket *sock,
+		     struct msghdr *msg, int size);
+int aa_inet_opt_perm(const char *op, u32 request, struct socket *sock, int level,
+		     int optname);
+int aa_inet_file_perm(const struct cred *subj_cred,
+		      struct aa_label *label, const char *op, u32 request,
+		      struct socket *sock);
+
+#endif /* __AA_AF_INET_H */
diff --git a/security/apparmor/include/net.h b/security/apparmor/include/net.h
index 0d0b0ce42723..2efc5417d1ba 100644
--- a/security/apparmor/include/net.h
+++ b/security/apparmor/include/net.h
@@ -102,13 +102,14 @@ int aa_af_perm(const struct cred *subj_cred, struct aa_label *label,
 static inline int aa_profile_af_sk_perm(struct aa_profile *profile,
 					struct apparmor_audit_data *ad,
 					u32 request,
-					struct sock *sk)
+					const struct sock *sk)
 {
 	return aa_profile_af_perm(profile, ad, request, sk->sk_family,
 				  sk->sk_type, sk->sk_protocol);
 }
-int aa_sk_perm(const char *op, u32 request, struct sock *sk);
-
+int aa_sk_perm(const char *op, u32 request, const struct sock *sk);
+int aa_label_sk_perm(const struct cred *subj_cred, struct aa_label *label,
+		     const char *op, u32 request, const struct sock *sk);
 int aa_sock_file_perm(const struct cred *subj_cred, struct aa_label *label,
 		      const char *op, u32 request,
 		      struct file *file);
diff --git a/security/apparmor/lsm.c b/security/apparmor/lsm.c
index ef5fb3b9c80a..65db1e5460bf 100644
--- a/security/apparmor/lsm.c
+++ b/security/apparmor/lsm.c
@@ -28,6 +28,7 @@
 #include <uapi/linux/lsm.h>
 
 #include "include/af_unix.h"
+#include "include/af_inet.h"
 #include "include/apparmor.h"
 #include "include/apparmorfs.h"
 #include "include/audit.h"
@@ -1266,13 +1267,21 @@ static int apparmor_socket_create(int family, int type, int protocol, int kern)
 
 	label = begin_current_label_crit_section(&needput);
 	if (!unconfined(label)) {
-		if (family == PF_UNIX)
+		switch (family) {
+		case PF_UNIX:
 			error = aa_unix_create_perm(label, family, type,
 						    protocol);
-		else
+			break;
+		case PF_INET:
+		case PF_INET6:
+			error = aa_inet_create_perm(label, family, type,
+						    protocol);
+			break;
+		default:
 			error = aa_af_perm(current_cred(), label, OP_CREATE,
 					   AA_MAY_CREATE, family, type,
 					   protocol);
+		}
 	}
 	end_current_label_crit_section(label, needput);
 
@@ -1367,8 +1376,13 @@ static int apparmor_socket_bind(struct socket *sock,
 	AA_BUG(!address);
 	AA_BUG(in_interrupt());
 
-	if (sock->sk->sk_family == PF_UNIX)
+	switch (sock->sk->sk_family) {
+	case PF_UNIX:
 		return aa_unix_bind_perm(sock, address, addrlen);
+	case PF_INET:
+	case PF_INET6:
+		return aa_inet_bind_perm(sock, address, addrlen);
+	}
 	return aa_sk_perm(OP_BIND, AA_MAY_BIND, sock->sk);
 }
 
@@ -1381,8 +1395,13 @@ static int apparmor_socket_connect(struct socket *sock,
 	AA_BUG(in_interrupt());
 
 	/* PF_UNIX goes through unix_stream_connect && unix_may_send */
-	if (sock->sk->sk_family == PF_UNIX)
+	switch (sock->sk->sk_family) {
+	case PF_UNIX:
 		return 0;
+	case PF_INET:
+	case PF_INET6:
+		return aa_inet_connect_perm(sock, address, addrlen);
+	}
 	return aa_sk_perm(OP_CONNECT, AA_MAY_CONNECT, sock->sk);
 }
 
@@ -1392,8 +1411,13 @@ static int apparmor_socket_listen(struct socket *sock, int backlog)
 	AA_BUG(!sock->sk);
 	AA_BUG(in_interrupt());
 
-	if (sock->sk->sk_family == PF_UNIX)
+	switch (sock->sk->sk_family) {
+	case PF_UNIX:
 		return aa_unix_listen_perm(sock, backlog);
+	case PF_INET:
+	case PF_INET6:
+		return aa_inet_listen_perm(sock, backlog);
+	}
 	return aa_sk_perm(OP_LISTEN, AA_MAY_LISTEN, sock->sk);
 }
 
@@ -1408,8 +1432,13 @@ static int apparmor_socket_accept(struct socket *sock, struct socket *newsock)
 	AA_BUG(!newsock);
 	AA_BUG(in_interrupt());
 
-	if (sock->sk->sk_family == PF_UNIX)
+	switch (sock->sk->sk_family) {
+	case PF_UNIX:
 		return aa_unix_accept_perm(sock, newsock);
+	case PF_INET:
+	case PF_INET6:
+		return aa_inet_accept_perm(sock, newsock);
+	}
 	return aa_sk_perm(OP_ACCEPT, AA_MAY_ACCEPT, sock->sk);
 }
 
@@ -1422,8 +1451,14 @@ static int aa_sock_msg_perm(const char *op, u32 request, struct socket *sock,
 	AA_BUG(in_interrupt());
 
 	/* PF_UNIX goes through unix_may_send */
-	if (sock->sk->sk_family == PF_UNIX)
+	switch (sock->sk->sk_family) {
+	case PF_UNIX:
 		return 0;
+	case PF_INET:
+	case PF_INET6:
+		return aa_inet_msg_perm(op, request, sock, msg, size);
+	}
+
 	return aa_sk_perm(op, request, sock->sk);
 }
 
@@ -1442,7 +1477,8 @@ static int apparmor_socket_sendmsg(struct socket *sock,
 	    (sk_is_tcp(sock->sk) ||
 	     (sk_is_inet(sock->sk) && sock->sk->sk_type == SOCK_STREAM &&
 	      sock->sk->sk_protocol == IPPROTO_MPTCP)))
-		error = aa_sk_perm(OP_CONNECT, AA_MAY_CONNECT, sock->sk);
+		error = aa_sock_msg_perm(OP_CONNECT, AA_MAY_CONNECT, sock,
+					 msg, size);
 
 	return error;
 }
@@ -1460,8 +1496,13 @@ static int aa_sock_perm(const char *op, u32 request, struct socket *sock)
 	AA_BUG(!sock->sk);
 	AA_BUG(in_interrupt());
 
-	if (sock->sk->sk_family == PF_UNIX)
+	switch (sock->sk->sk_family) {
+	case PF_UNIX:
 		return aa_unix_sock_perm(op, request, sock);
+	case PF_INET:
+	case PF_INET6:
+		return aa_inet_sock_perm(op, request, sock);
+	}
 	return aa_sk_perm(op, request, sock->sk);
 }
 
@@ -1483,8 +1524,13 @@ static int aa_sock_opt_perm(const char *op, u32 request, struct socket *sock,
 	AA_BUG(!sock->sk);
 	AA_BUG(in_interrupt());
 
-	if (sock->sk->sk_family == PF_UNIX)
+	switch (sock->sk->sk_family) {
+	case PF_UNIX:
 		return aa_unix_opt_perm(op, request, sock, level, optname);
+	case PF_INET:
+	case PF_INET6:
+		return aa_inet_opt_perm(op, request, sock, level, optname);
+	}
 	return aa_sk_perm(op, request, sock->sk);
 }
 
diff --git a/security/apparmor/net.c b/security/apparmor/net.c
index e3a3708a057e..678d1417efd6 100644
--- a/security/apparmor/net.c
+++ b/security/apparmor/net.c
@@ -9,6 +9,7 @@
  */
 
 #include "include/af_unix.h"
+#include "include/af_inet.h"
 #include "include/apparmor.h"
 #include "include/audit.h"
 #include "include/cred.h"
@@ -133,12 +134,12 @@ void audit_net_cb(struct audit_buffer *ab, void *va)
 	audit_log_format(ab, " protocol=%d", ad->net.protocol);
 
 	if (ad->request & NET_PERMS_MASK) {
-		audit_log_format(ab, " requested_mask=");
+		audit_log_format(ab, " requested=");
 		aa_audit_perm_mask(ab, ad->request, NULL, 0,
 				   net_mask_names, NET_PERMS_MASK);
 
 		if (ad->denied & NET_PERMS_MASK) {
-			audit_log_format(ab, " denied_mask=");
+			audit_log_format(ab, " denied=");
 			aa_audit_perm_mask(ab, ad->denied, NULL, 0,
 					   net_mask_names, NET_PERMS_MASK);
 		}
@@ -282,10 +283,8 @@ int aa_af_perm(const struct cred *subj_cred, struct aa_label *label,
 					   type, protocol));
 }
 
-static int aa_label_sk_perm(const struct cred *subj_cred,
-			    struct aa_label *label,
-			    const char *op, u32 request,
-			    struct sock *sk)
+int aa_label_sk_perm(const struct cred *subj_cred, struct aa_label *label,
+		     const char *op, u32 request, const struct sock *sk)
 {
 	struct aa_sk_ctx *ctx = aa_sock(sk);
 	int error = 0;
@@ -305,7 +304,7 @@ static int aa_label_sk_perm(const struct cred *subj_cred,
 	return error;
 }
 
-int aa_sk_perm(const char *op, u32 request, struct sock *sk)
+int aa_sk_perm(const char *op, u32 request, const struct sock *sk)
 {
 	struct aa_label *label;
 	bool needput;
@@ -334,8 +333,13 @@ int aa_sock_file_perm(const struct cred *subj_cred, struct aa_label *label,
 	if (!sock || !sock->sk)
 		return 0;
 
-	if (sock->sk->sk_family == PF_UNIX)
+	switch (sock->sk->sk_family) {
+	case PF_UNIX:
 		return aa_unix_file_perm(subj_cred, label, op, request, file);
+	case PF_INET:
+	case PF_INET6:
+		return aa_inet_file_perm(subj_cred, label, op, request, sock);
+	}
 	return aa_label_sk_perm(subj_cred, label, op, request, sock->sk);
 }
 
-- 
2.55.0

